smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php

```diff
@@ -0,0 +1,277 @@
+      '#options' => $autoresponder_list,
+      '#default_value' => $config->get('smaily_autoresponder', 1),
+      '#empty_option' => $this->t('Select autoresponder'),
+    ];
+
+    $form['smaily_container']['smaily_button_title'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Subscribe button title'),
+      '#default_value' => $config->get('smaily_button_title', 'Subscribe to newsletter'),
+      '#required' => TRUE,
+    ];
+
+    $form['smaily_container']['submit'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Save'),
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    // Validation handled by #required and in ajax function.
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    // Normalize subdomain.
+    // First, try to parse as full URL.
+    // If that fails, try to parse as subdomain.sendsmaily.net, and, then clean up subdomain and pass as is.
+    $subdomain = trim($form_state->getValue('smaily_domain'));
+    if (filter_var($subdomain, FILTER_VALIDATE_URL)) {
+      $url = parse_url($subdomain);
+      $parts = explode('.', $url['host']);
+      $subdomain = count($parts) >= 3 ? $parts[0] : '';
+    }
+    elseif (preg_match('/^[^\.]+\.sendsmaily\.net$/', $subdomain)) {
+      $parts = explode('.', $subdomain);
+      $subdomain = $parts[0];
+    }
+    $subdomain = preg_replace('/[^a-zA-Z0-9]+/', '', $subdomain);
+    // Save form to config.
+    $this->config('smaily_for_drupal.adminsettings')
+      ->set('smaily_domain', $subdomain)
+      ->set('smaily_username', trim($form_state->getValue('smaily_username')))
+      ->set('smaily_password', trim($form_state->getValue('smaily_password')))
+      ->set('smaily_autoresponder', trim($form_state->getValue('smaily_autoresponder')))
+      ->set('smaily_button_title', trim($form_state->getValue('smaily_button_title')))
+      ->save();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateCredentials(array &$form, FormStateInterface $form_state) {
+    $ajax_response = new AjaxResponse();
+
+    $username = trim($form_state->getValue('smaily_username'));
+    $password = trim($form_state->getValue('smaily_password'));
+    $subdomain = trim($form_state->getValue('smaily_domain'));
+    // Drupal's validation is only done with submit.
+    if (empty($username) || empty($password) || empty($subdomain)) {
+      $ajax_response->addCommand(
+        new HtmlCommand(
+          '.smaily_message',
+          '<div class="messages messages--error">' . 'Please fill out all fields.' . '</div>')
+      );
+      return $ajax_response;
+    }
+
+    // Normalize subdomain.
+    // First, try to parse as full URL.
+    // If that fails, try to parse as subdomain.sendsmaily.net, and, then clean up subdomain and pass as is.
+    if (filter_var($subdomain, FILTER_VALIDATE_URL)) {
```

Comment:
Please extract this functionality into a separate method that takes the unformatted subdomain as its value and returns formatted subdomain.